hedera-state-validator/src/main/java/com/hedera/statevalidation/poc/ChunkedFileIterator.java

```diff
@@ -0,0 +1,251 @@
+    private long currentDataItemFilePosition;
+    private boolean closed = false;
+
+    private long boundaryOffset = 0L;
+
+    public ChunkedFileIterator(
+            Path path, DataFileMetadata metadata, Type dataType, long startByte, long endByte,
+            AtomicLong totalBoundarySearchMillis)
+            throws IOException {
+        this.channel = FileChannel.open(path, StandardOpenOption.READ);
+        this.metadata = metadata;
+
+        this.startByte = startByte;
+        this.endByte = endByte;
+
+        this.dataType = dataType;
+
+        if (startByte > 0) {
+            // Find boundary, then position channel and open streams
+            long startTime = System.currentTimeMillis();
+            this.boundaryOffset = findBoundaryOffset();
+            long boundaryOffsetSearchTime = System.currentTimeMillis() - startTime;
+//            System.out.println("Found boundary offset in:" + boundaryOffsetSearchTime + " ms");
+            totalBoundarySearchMillis.addAndGet(boundaryOffsetSearchTime);
+            channel.position(startByte + boundaryOffset);
+            openStreams();
+        } else {
+            // At file start
+            channel.position(startByte);
+            openStreams();
+        }
+    }
+
+    private void openStreams() {
+        var channelStream = Channels.newInputStream(channel);
+        // 16777216 took from MerkleDbConfig.iteratorInputBufferBytes
+        this.bufferedInputStream = new BufferedInputStream(channelStream, 16777216);
+        this.in = new ReadableStreamingData(bufferedInputStream);
+    }
+
+    private long findBoundaryOffset() throws IOException {
+        // Use buffer to minimize disk I/O and channel repositioning
+        // It should account for boundary + full data item to validate its proto schema
+        // 16777216 took from MerkleDbConfig.iteratorInputBufferBytes
+        ByteBuffer scanBuffer = ByteBuffer.allocate(16777216);
+
+        // Read large chunk at current position
+        scanBuffer.clear();
+        channel.position(startByte);
+        int bytesRead = channel.read(scanBuffer);
+        if (bytesRead <= 0) {
+            throw new IOException("No valid data item boundary found in chunk");
+        }
+
+        scanBuffer.flip();
+        BufferedData bufferData = BufferedData.wrap(scanBuffer);
+
+        // Scan through buffer looking for valid boundary
+        while (bufferData.hasRemaining()) {
+            long positionInBuffer = bufferData.position();
+
+            try {
+                int tag = bufferData.readVarInt(false);
+                int fieldNum = tag >> TAG_FIELD_OFFSET;
+
+                if (fieldNum == FIELD_DATAFILE_ITEMS.number()) {
+                    int dataItemSize = bufferData.readVarInt(false);
+                    long dataStartPosition = bufferData.position();
+
+                    if (dataItemSize > 0) {
+                        bufferData.limit(dataStartPosition + dataItemSize);
```

Comment:
Data item size should be checked against buffer limits

Reply:
Agree, addressed